### PR TITLE
Added more options to Java client

### PR DESCRIPTION
**Pull request by pa314159**
All changes are in the README

### CODE REVIEW

File: boat-maven-plugin/src/main/java/com/backbase/oss/boat/GenerateMojo.java [hunk 1]
```diff
@@ -73,6 +77,16 @@
 @Slf4j
 public class GenerateMojo extends AbstractMojo {
 
+    private static String trimCSV(String text) {
```
Comment (bartveenstra) [inline boat-maven-plugin/src/main/java/com/backbase/oss/boat/GenerateMojo.java:80]: shouldn't this be at the bottom of the class?

File: boat-scaffold/src/main/java/com/backbase/oss/codegen/java/BoatJavaCodeGen.java [hunk 2]
```diff
@@ -1,16 +1,165 @@
...
+        }
+        writePropertyBack(USE_WITH_MODIFIERS, this.useWithModifiers);
+
+        if (this.additionalProperties.containsKey(USE_SET_FOR_UNIQUE_ITEMS)) {
+            this.useSetForUniqueItems = convertPropertyToBoolean(USE_SET_FOR_UNIQUE_ITEMS);
+        }
+        writePropertyBack(USE_SET_FOR_UNIQUE_ITEMS, this.useSetForUniqueItems);
+
+        if (this.useSetForUniqueItems) {
+            this.typeMapping.put("set", "java.util.Set");
```
Comment (bartveenstra) [inline boat-scaffold/src/main/java/com/backbase/oss/codegen/java/BoatJavaCodeGen.java:95]: Sonar will probably tell you to use contants here :)

File: boat-scaffold/src/main/templates/boat-java/libraries/resttemplate/ApiClient.mustache [hunk 3]
```diff
@@ -42,6 +42,16 @@ import com.fasterxml.jackson.databind.ObjectMapper;
 import org.openapitools.jackson.nullable.JsonNullableModule;
 {{/threetenbp}}
 
+{{#serviceId}}
+import static com.backbase.buildingblocks.backend.communication.http.HttpCommunicationConfiguration.INTER_SERVICE_REST_TEMPLATE_BEAN_NAME;
```
Comment (bartveenstra) [inline boat-scaffold/src/main/templates/boat-java/libraries/resttemplate/ApiClient.mustache:46]: I actually use an other rest template in my projects. This is too tied to the Backbase SSDK and interservcie comms. 

@jjjasper @torrespro  Your thoughts?

File: boat-scaffold/src/main/templates/boat-java/libraries/resttemplate/ApiClient.mustache [hunk 4]
```diff
@@ -96,42 +108,60 @@ public class ApiClient {
...
     }
+{{/serviceId}}
+{{#useJacksonConversion}}
+    @Autowired
+    private ObjectMapper objectMapper;
+{{/useJacksonConversion}}
 
     @Autowired
-    public ApiClient(RestTemplate restTemplate) {
+    public ApiClient({{#serviceId}}@Qualifier(INTER_SERVICE_REST_TEMPLATE_BEAN_NAME) {{/serviceId}}RestTemplate restTemplate) {
```
Comment (bartveenstra) [inline boat-scaffold/src/main/templates/boat-java/libraries/resttemplate/ApiClient.mustache:133]: Too specific. You can externalise qualifier to a variable to so devs can choose which rest template to inject.

**Review by bartveenstra — CHANGES_REQUESTED** [review]

File: README.md [hunk 5]
```diff
@@ -19,8 +19,19 @@ The project is very much Work In Progress and will be published on maven central
...
+## 0.11.0
+
+* *Maven Plugin*
+   * Added `apisToGenerate` mojo parameter to `boat:generate` to filter out generated APIs.
+
+* *Java Generator*
+  * Added `useWithModifiers` option to use the `with` prefix for POJO modifiers (defaults to `false`).
+  * Added `useClassLevelBeanValidation` option (defaults to `false`).
+  * Added `useJacksonConversion` to use Jackson for parameters conversion instead of `toString` (defaults to `false`).
+  * Added `backbaseServiceId` to use generate API bean factories (defaults to `""`, so no factory is generated).
```
[the remark this answers is not in the document]
Comment (pa314159) [reply README.md]: it's actually `serviceId`...
however, I'll come with a more neutral approach for ApiClientConfiguration

File: README.md [hunk 6]
```diff
@@ -19,8 +19,57 @@ The project is very much Work In Progress and will be published on maven central
...
+
+* *Java Generator (resttemplate only)*
+  * Added `useWithModifiers` option to use the `with` prefix for POJO modifiers (defaults to `false`).
+  * Added `useSetForUniqueItems` to map arrays containing `uniqueItems` to `Set` (defaults to `false`).
+  * Added `useClassLevelBeanValidation` option (defaults to `false`).
+  * Added `useJacksonConversion` to use Jackson for parameters' conversion instead of `toString` (defaults to `false`). Enabling this option aligns the formats of the query parameters with JSON values, e.g. OfssetDateTime is passed correctly as RFC3339.
+  * Added `serviceId` to generate capability specific API factories (defaults to `""`, so no factory is generated).
+  * Added `configName` as the configuration name under `backbase.communication.services.` (defaults to `serviceId`).
+ 
+** *Example of Access Control client configuration*
```
[the remark this answers is not in the document]
Comment (pa314159) [reply README.md:36]: Petstore is too schematic and it generates only one interface. I'll look for something else.

File: boat-maven-plugin/src/main/java/com/backbase/oss/boat/GenerateMojo.java (hunk 1 above)
Comment (pa314159) [reply boat-maven-plugin/src/main/java/com/backbase/oss/boat/GenerateMojo.java:80]: It's private static shouldn't be at the top?

File: boat-scaffold/src/main/templates/boat-java/libraries/resttemplate/ApiClient.mustache (hunk 3 above)
Comment (pa314159) [reply boat-scaffold/src/main/templates/boat-java/libraries/resttemplate/ApiClient.mustache:46]: Agreed, this is taken from Transactions I think... Can you point me to your template?

**Comment by pa314159** [comment]
I withdrew the the ApiClientConfiguration template.
The client configuration is too project specific, so such a template would end up by not being used at all :)

File: boat-scaffold/src/main/templates/boat-java/libraries/resttemplate/ApiClient.mustache (hunk 4 above)
Comment (pa314159) [reply boat-scaffold/src/main/templates/boat-java/libraries/resttemplate/ApiClient.mustache]: The qualifier is now an option

File: README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (pa314159) [reply README.md]: removed

File: boat-scaffold/src/main/templates/boat-java/libraries/resttemplate/ApiClient.mustache [hunk 7]
```diff
@@ -96,42 +108,60 @@ public class ApiClient {
...
     }
+{{/backbaseServiceId}}
+{{#useJacksonConversion}}
+    @Autowired
+    private ObjectMapper objectMapper;
+{{/useJacksonConversion}}
 
     @Autowired
-    public ApiClient(RestTemplate restTemplate) {
+    public ApiClient({{#backbaseServiceId}}@Qualifier(INTER_SERVICE_REST_TEMPLATE_BEAN_NAME) {{/backbaseServiceId}}RestTemplate restTemplate) {
```
[the remark this answers is not in the document]
Comment (pa314159) [reply boat-scaffold/src/main/templates/boat-java/libraries/resttemplate/ApiClient.mustache]: changed to {{restTemplateBeanName}}

File: boat-scaffold/src/main/java/com/backbase/oss/codegen/java/BoatJavaCodeGen.java (hunk 2 above)
Comment (pa314159) [reply boat-scaffold/src/main/java/com/backbase/oss/codegen/java/BoatJavaCodeGen.java]: done

File: boat-scaffold/src/main/templates/boat-java/libraries/resttemplate/ApiClient.mustache (hunk 3 above)
Comment (pa314159) [reply boat-scaffold/src/main/templates/boat-java/libraries/resttemplate/ApiClient.mustache]: removed

**Review by bartveenstra — APPROVED** [review]

**Pull request merged**